src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,486 @@
+use serde::{de::Deserializer, Deserialize, Serialize};
+
+use super::VerifyToken;
+use crate::tokenserver::{
+    error::{ErrorLocation, TokenserverError},
+    settings::Settings,
+};
+
+use core::time::Duration;
+use std::{convert::TryFrom, sync::Arc};
+
+/// The information extracted from a valid BrowserID assertion.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+pub struct VerifyOutput {
+    pub device_id: Option<String>,
+    pub email: String,
+    pub generation: Option<i64>,
+    pub keys_changed_at: Option<i64>,
+}
+
+/// The verifier used to verify BrowserID assertions.
+#[derive(Clone)]
+pub struct RemoteVerifier {
+    audience: String,
+    issuer: String,
+    fxa_verifier_url: String,
+    // We need to use an Arc here because reqwest's blocking client doesn't use one internally,
+    // and we want to share keep-alive connections across threads
+    request_client: Arc<ReqwestClient>,
+}
+
+impl TryFrom<&Settings> for RemoteVerifier {
+    type Error = Error;
+
+    fn try_from(settings: &Settings) -> Result<Self, Error> {
+        Ok(Self {
+            audience: settings.fxa_browserid_audience.clone(),
+            issuer: settings.fxa_browserid_issuer.clone(),
+            request_client: Arc::new(
+                ReqwestClient::builder()
+                    .timeout(Duration::new(settings.fxa_browserid_request_timeout, 0))
+                    .build()
+                    .map_err(|_| Error::from(()))?,
+            ),
+            fxa_verifier_url: settings.fxa_browserid_server_url.clone(),
+        })
+    }
+}
+
+#[async_trait]
+impl VerifyToken for RemoteVerifier {
+    type Output = VerifyOutput;
+
+    /// Verifies a BrowserID assertion. Returns `VerifyOutput` for valid assertions and a
+    /// `TokenserverError` for invalid assertions.
+    async fn verify(&self, assertion: String) -> Result<VerifyOutput, TokenserverError> {
+        let verifier = self.clone();
+
+        web::block(move || {
+            let response = verifier
+                .request_client
+                .post(&verifier.fxa_verifier_url)
+                .json(&VerifyRequest {
+                    assertion,
+                    audience: verifier.audience.clone(),
+                    trusted_issuers: [verifier.issuer.clone()],
+                })
+                .send()
+                .map_err(|e| {
+                    if e.is_connect() {
+                        // If we are unable to reach the FxA server, report a 503 to the client
+                        TokenserverError::resource_unavailable()
+                    } else {
+                        // If any other error occurs during the request, report a 401 to the client
+                        TokenserverError::invalid_credentials("Unauthorized")
+                    }
+                })?;
+
+            // If FxA responds with an HTTP status other than 200, report a 503 to the client
+            if response.status() != StatusCode::OK {
```

Comment:
nit: You might be able to kill this block entirely by using `error_for_status` along with including `e.is_status()` along w/ the `if e.is_connect` check.